runners/extensions-java/metrics/src/test/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSinkTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.extensions.metrics;
+
+import static org.junit.Assert.assertTrue;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/** Test class for MetricsGraphiteSink. */
+public class MetricsGraphiteSinkTest {
+  private static NetworkMockServer graphiteServer;
+  private static int port;
+
+  @BeforeClass
+  public static void beforeClass() throws IOException, InterruptedException {
+    // get free local port
+    ServerSocket serverSocket = new ServerSocket(0);
+    port = serverSocket.getLocalPort();
+    serverSocket.close();
+    graphiteServer = new NetworkMockServer(port);
+    Thread.sleep(200);
+    graphiteServer.clear();
+    graphiteServer.start();
+  }
+
+  @Before
+  public void before() {
+    graphiteServer.clear();
+  }
+
+  @AfterClass
+  public static void afterClass() throws IOException {
+    graphiteServer.stop();
+  }
+
+  @Test
+  public void testWriteMetricsWithCommittedSupported() throws Exception {
+    MetricQueryResults metricQueryResults = new CustomMetricQueryResults(true);
+    PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    pipelineOptions.setMetricsGraphitePort(port);
+    pipelineOptions.setMetricsGraphiteHost("127.0.0.1");
+    MetricsGraphiteSink metricsGraphiteSink = new MetricsGraphiteSink(pipelineOptions);
+    metricsGraphiteSink.writeMetrics(metricQueryResults);
+    Thread.sleep(2000L);
```

Comment:
Yes because, when we write messages to the socket, then on a different thread `NetworkMockServer` reads the socket and adds messages to an arrayList so that they could be read in the assert. On heavy load jenkins server, the test might fail because the `NetworkMockServer` does not have enough time to add the messages to the arraylist

Reply:
Well, potentially it's still can be a cause of flaky tests - we never know how much time you need to wait. Could you avoid using two threads and socket in this case? Or use approach based on `CountDownLatch`?

Reply:
+1